File: src/ai_council/strategies/redteam.py
"""Red Team — adversarial critique and defense."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass

from ai_council.config import get_aggregator, get_proposers
from ai_council.prompts import (
    REDTEAM_ATTACKER_SYSTEM,
    REDTEAM_ATTACKER_TEMPLATE,
    REDTEAM_DEFENSE_SYSTEM,
    REDTEAM_DEFENSE_TEMPLATE,
    REDTEAM_JUDGE_SYSTEM,
    REDTEAM_JUDGE_TEMPLATE,
    format_proposals,
)
from ai_council.providers import ModelResponse, call_model, call_models_parallel
from ai_council.scoring import score_agreement


@dataclass
class RedTeamResult:
    proposals: list[ModelResponse]
    initial_attack: ModelResponse
    defenses: list[ModelResponse]
    targeted_attack: ModelResponse
    synthesis: ModelResponse
    total_ms: int = 0
    agreement_score: int | None = None
    agreement_reason: str | None = None


def _select_attacker(prompt: str, proposers: list) -> int:
    """Select attacker index via prompt hash. Deterministic per prompt."""
    return hash(prompt) % len(proposers)


async def run_redteam(prompt: str) -> RedTeamResult:
    all_proposers = get_proposers()

    if len(all_proposers) < 2:
        from ai_council.strategies.moa import run_moa
        moa = await run_moa(prompt)
        return RedTeamResult(
            proposals=moa.proposals, initial_attack=moa.synthesis,
            defenses=[], targeted_attack=moa.synthesis,
            synthesis=moa.synthesis, total_ms=moa.total_ms,
        )

    attacker_idx = _select_attacker(prompt, all_proposers)
    attacker_cfg = all_proposers[attacker_idx]
    proposer_cfgs = [p for i, p in enumerate(all_proposers) if i != attacker_idx]

    total = 0

    # Round 1: Proposers answer
    proposals = await call_models_parallel(
        proposer_cfgs, [{"role": "user", "content": prompt}]
    )
    total += sum(p.latency_ms for p in proposals)

    ok_proposals = [p for p in proposals if p.succeeded]
    if not ok_proposals:
        fail = ModelResponse("", "redteam", "All proposers failed.", 0, "no proposals")
        return RedTeamResult(
            proposals=proposals, initial_attack=fail,
            defenses=[], targeted_attack=fail, synthesis=fail, total_ms=total,
        )

    # Round 1: Attacker critiques
    attack_prompt = REDTEAM_ATTACKER_TEMPLATE.format(
        prompt=prompt, proposals=format_proposals(ok_proposals)
    )
    initial_attack = await call_model(
        attacker_cfg,
        [{"role": "system", "content": REDTEAM_ATTACKER_SYSTEM},
         {"role": "user", "content": attack_prompt}],
    )
    total += initial_attack.latency_ms

    # Round 2: Proposers defend
    defense_tasks = []
    for cfg in proposer_cfgs:
        own = next((p for p in ok_proposals if p.model == cfg.model), None)
        if not own:
            continue
        defense_prompt = REDTEAM_DEFENSE_TEMPLATE.format(
            prompt=prompt, own_response=own.content,
            attack=initial_attack.content if initial_attack.succeeded else "No attack available.",
        )
        defense_tasks.append(
            call_model(cfg, [
                {"role": "system", "content": REDTEAM_DEFENSE_SYSTEM},
                {"role": "user", "content": defense_prompt},
            ])
        )
    defenses = list(await asyncio.gather(*defense_tasks))
    total += sum(d.latency_ms for d in defenses)

    # Round 2: Attacker targets defenses
    ok_defenses = [d for d in defenses if d.succeeded]
    targeted_prompt = REDTEAM_ATTACKER_TEMPLATE.format(
        prompt=prompt,
        proposals=format_proposals(ok_defenses) if ok_defenses else "No defenses.",
    )
    targeted_attack = await call_model(
        attacker_cfg,
        [{"role": "system", "content": REDTEAM_ATTACKER_SYSTEM},
         {"role": "user", "content": targeted_prompt}],
    )
    total += targeted_attack.latency_ms

    # Final: Judge synthesizes hardened answer
    judge_prompt = REDTEAM_JUDGE_TEMPLATE.format(
        prompt=prompt,
        proposals=format_proposals(ok_proposals),
        attack=initial_attack.content if initial_attack.succeeded else "N/A",
        defenses=format_proposals(ok_defenses) if ok_defenses else "N/A",
        targeted_attack=targeted_attack.content if targeted_attack.succeeded else "N/A",
    )
    synthesis = await call_model(
        get_aggregator(),
        [{"role": "system", "content": REDTEAM_JUDGE_SYSTEM},
         {"role": "user", "content": judge_prompt}],
    )
    total += synthesis.latency_ms

    # Score agreement on defenses (or proposals if no defenses)
    score_targets = ok_defenses if ok_defenses else ok_proposals
    score, reason = await score_agreement(score_targets, prompt)

    return RedTeamResult(
        proposals=proposals, initial_attack=initial_attack,
        defenses=defenses, targeted_attack=targeted_attack,
        synthesis=synthesis, total_ms=total,
        agreement_score=score, agreement_reason=reason,
    )

File: internal/strategy/redteam.go
package strategy

import (
	"context"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"strings"
	"time"

	"github.com/avicuna/ai-council-personal/internal/prompt"
	"github.com/avicuna/ai-council-personal/internal/provider"
	"github.com/avicuna/ai-council-personal/internal/scoring"
)

// RedTeam implements an adversarial red team strategy.
// One provider is selected as the attacker to critique others' proposals.
// The flow is: proposals → critique → defenses → targeted attack → judge synthesis.
type RedTeam struct{}

// NewRedTeam creates a new RedTeam strategy instance.
func NewRedTeam() *RedTeam {
	return &RedTeam{}
}

// Execute runs the red team strategy:
// 1. If <2 providers: fall back to MoA
// 2. Select attacker deterministically using sha256(prompt) % len(proposers)
// 3. Defenders propose initial answers
// 4. Attacker critiques proposals
// 5. Defenders revise and defend
// 6. Attacker targets defenses
// 7. Judge synthesizes hardened answer
// 8. Score agreement on defenses
func (r *RedTeam) Execute(ctx context.Context, opts *Options) (*Result, error) {
	if len(opts.Proposers) == 0 {
		return nil, fmt.Errorf("no proposers configured")
	}

	// Fall back to MoA if fewer than 2 providers (need attacker + at least 1 defender)
	if len(opts.Proposers) < 2 {
		moa := NewMoA()
		return moa.Execute(ctx, opts)
	}

	start := time.Now()

	// Select attacker deterministically based on prompt
	attackerIdx := selectAttacker(opts.Request.UserPrompt, len(opts.Proposers))
	attacker := opts.Proposers[attackerIdx]

	// Separate defenders from attacker
	var defenders []provider.Provider
	for i, p := range opts.Proposers {
		if i != attackerIdx {
			defenders = append(defenders, p)
		}
	}

	// Round 1: Defenders propose initial answers
	proposalResults := provider.QueryAll(ctx, defenders, opts.Request, opts.Progress)
	if len(proposalResults.Responses) == 0 {
		return nil, fmt.Errorf("all defenders failed: %v", proposalResults.Errors)
	}

	initialProposals := proposalResults.Responses

	// Attacker critiques proposals
	attackPrompt := buildAttackPrompt(opts.Request.UserPrompt, initialProposals)
	attackResp, err := attacker.Query(ctx, attackPrompt)
	if err != nil {
		return nil, fmt.Errorf("attacker failed: %w", err)
	}

	// Send progress event for attacker
	if opts.Progress != nil {
		select {
		case opts.Progress <- provider.ProgressEvent{
			Model:   attacker.Name(),
			Status:  "critique",
			Latency: time.Duration(attackResp.LatencyMs) * time.Millisecond,
		}:
		case <-ctx.Done():
		}
	}

	// Round 2: Defenders revise addressing the attack
	var defenseProviders []provider.Provider
	var defenseRequests []*provider.Request

	for _, proposal := range initialProposals {
		// Find the original defender
		var defender provider.Provider
		for _, d := range defenders {
			if d.Name() == proposal.Name {
				defender = d
				break
			}
		}
		if defender == nil {
			continue
		}

		defenseReq := buildDefenseRequest(opts.Request.UserPrompt, proposal.Content, attackResp.Content)
		defenseProviders = append(defenseProviders, defender)
		defenseRequests = append(defenseRequests, defenseReq)
	}

	defenseResults := queryAllWithRequests(ctx, defenseProviders, defenseRequests, opts.Progress)
	if len(defenseResults.Responses) == 0 {
		return nil, fmt.Errorf("all defenders failed to defend: %v", defenseResults.Errors)
	}

	defenses := defenseResults.Responses

	// Attacker targets defenses
	targetedAttackPrompt := buildTargetedAttackPrompt(opts.Request.UserPrompt, initialProposals, attackResp.Content, defenses)
	targetedAttackResp, err := attacker.Query(ctx, targetedAttackPrompt)
	if err != nil {
		// Non-fatal: proceed without targeted attack
		targetedAttackResp = &provider.Response{
			Content: "(targeted attack failed)",
		}
	}

	// Send progress event for targeted attack
	if opts.Progress != nil {
		select {
		case opts.Progress <- provider.ProgressEvent{
			Model:   attacker.Name(),
			Status:  "targeted_critique",
			Latency: time.Duration(targetedAttackResp.LatencyMs) * time.Millisecond,
		}:
		case <-ctx.Done():
		}
	}

	// Judge synthesizes hardened answer
	judgePrompt := buildJudgePrompt(opts.Request.UserPrompt, initialProposals, attackResp.Content, defenses, targetedAttackResp.Content)
	synthesis, err := opts.Aggregator.Query(ctx, judgePrompt)
	if err != nil {
		return nil, fmt.Errorf("judge failed: %w", err)
	}

	// Convert defenses to pointer slice for Result
	proposals := make([]*provider.Response, len(defenses))
	for i := range defenses {
		proposals[i] = &defenses[i]
	}

	// Score agreement on defenses if scorer provided
	var agreementScore *int
	var agreementReason string

	if opts.Scorer != nil {
		score, err := scoring.ScoreAgreement(ctx, opts.Scorer, defenses, opts.Request.UserPrompt)
		if err != nil {
			agreementReason = fmt.Sprintf("scoring failed: %v", err)
		} else if score != nil {
			agreementScore = &score.Score
			agreementReason = score.Reason
		}
	}

	totalMs := time.Since(start).Milliseconds()
	return &Result{
		Proposals:       proposals,
		Synthesis:       synthesis,
		Rounds:          2, // Red team is always 2 rounds (proposals + defenses)
		AgreementScore:  agreementScore,
		AgreementReason: agreementReason,
		TotalMs:         totalMs,
	}, nil
}

// selectAttacker selects an attacker index deterministically based on prompt.
func selectAttacker(prompt string, numProviders int) int {
	hash := sha256.Sum256([]byte(prompt))
	// Use first 8 bytes as uint64
	hashVal := binary.BigEndian.Uint64(hash[:8])
	return int(hashVal % uint64(numProviders))
}

// buildAttackPrompt constructs the attacker's critique prompt.
func buildAttackPrompt(userPrompt string, proposals []provider.Response) *provider.Request {
	proposalsText := prompt.FormatProposals(proposals)
	attackPrompt := strings.ReplaceAll(prompt.RedTeamAttackerTemplate, "{prompt}", userPrompt)
	attackPrompt = strings.ReplaceAll(attackPrompt, "{proposals}", proposalsText)

	return &provider.Request{
		SystemPrompt: prompt.RedTeamAttackerSystem,
		UserPrompt:   attackPrompt,
		Temperature:  nil,
		MaxTokens:    2000,
	}
}

// buildDefenseRequest constructs a defender's defense prompt.
func buildDefenseRequest(userPrompt, ownResponse, attack string) *provider.Request {
	defensePrompt := strings.ReplaceAll(prompt.RedTeamDefenseTemplate, "{prompt}", userPrompt)
	defensePrompt = strings.ReplaceAll(defensePrompt, "{own_response}", ownResponse)
	defensePrompt = strings.ReplaceAll(defensePrompt, "{attack}", attack)

	return &provider.Request{
		SystemPrompt: prompt.RedTeamDefenseSystem,
		UserPrompt:   defensePrompt,
		Temperature:  nil,
		MaxTokens:    3000,
	}
}

// buildTargetedAttackPrompt constructs the attacker's targeted critique of defenses.
func buildTargetedAttackPrompt(userPrompt string, proposals []provider.Response, initialAttack string, defenses []provider.Response) *provider.Request {
	proposalsText := prompt.FormatProposals(proposals)
	defensesText := prompt.FormatProposals(defenses)

	// Use the defense template but adapt it for targeted attack
	targetedPrompt := fmt.Sprintf("Original question: %s\n\nOriginal proposals:\n%s\n\nYour initial critique:\n%s\n\nDefenses:\n%s\n\nProvide a targeted critique addressing the defenses.",
		userPrompt, proposalsText, initialAttack, defensesText)

	return &provider.Request{
		SystemPrompt: prompt.RedTeamAttackerSystem,
		UserPrompt:   targetedPrompt,
		Temperature:  nil,
		MaxTokens:    2000,
	}
}

// buildJudgePrompt constructs the judge's synthesis prompt.
func buildJudgePrompt(userPrompt string, proposals []provider.Response, attack string, defenses []provider.Response, targetedAttack string) *provider.Request {
	proposalsText := prompt.FormatProposals(proposals)
	defensesText := prompt.FormatProposals(defenses)

	judgePrompt := strings.ReplaceAll(prompt.RedTeamJudgeTemplate, "{prompt}", userPrompt)
	judgePrompt = strings.ReplaceAll(judgePrompt, "{proposals}", proposalsText)
	judgePrompt = strings.ReplaceAll(judgePrompt, "{attack}", attack)
	judgePrompt = strings.ReplaceAll(judgePrompt, "{defenses}", defensesText)
	judgePrompt = strings.ReplaceAll(judgePrompt, "{targeted_attack}", targetedAttack)

	return &provider.Request{
		SystemPrompt: prompt.RedTeamJudgeSystem,
		UserPrompt:   judgePrompt,
		Temperature:  nil,
		MaxTokens:    4000,
	}
}
